README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# isogeochem: <img src="man/figures/logo.png" align="right" width="140"/> <br/> Tools for Stable Isotope Geochemistry

<!-- badges: start -->

[![R-CMD-check](https://github.com/davidbajnai/isogeochem/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/davidbajnai/isogeochem/actions/workflows/R-CMD-check.yaml)
[![codecov](https://codecov.io/gh/davidbajnai/isogeochem/branch/main/graph/badge.svg?token=srtusVq2Im)](https://app.codecov.io/gh/davidbajnai/isogeochem)
[![CodeFactor](https://www.codefactor.io/repository/github/davidbajnai/isogeochem/badge)](https://www.codefactor.io/repository/github/davidbajnai/isogeochem)
[![repostatus](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
[![CRAN
status](https://www.r-pkg.org/badges/version/isogeochem)](https://CRAN.R-project.org/package=isogeochem)
[![CRAN
count](https://cranlogs.r-pkg.org/badges/grand-total/isogeochem)](https://cranlogs.r-pkg.org/badges/grand-total/isogeochem)
[![DOI](https://zenodo.org/badge/401782303.svg)](https://zenodo.org/badge/latestdoi/401782303)
<!-- badges: end -->

`isogeochem` makes working with stable oxygen, carbon, and clumped
isotope data straightforward and reproducible. It offers tools to
quickly calculate:

- carbonate *δ*<sup>18</sup>O, *δ*<sup>17</sup>O, *∆*<sub>47</sub>, and
  *∆*<sub>48</sub> values at a given temperature
- carbonate growth temperatures from *δ*<sup>18</sup>O,
  *∆*<sub>47</sub>, and *∆*<sub>48</sub> values
- isotope fractionation factors, e.g., carbonate/water, quartz/water
- model DIC speciation as a function of temperature, pH, and salinity
- convert between the VSMOW and VPDB scales

The list of available proxy–temperature calibrations is growing with
each new released version. </br> <b>Please get in touch if you have
suggestions to include!</b>

## Getting started

### Installation

Install the released version of `isogeochem` from CRAN.

``` r
install.packages("isogeochem")
```

Install the development version of `isogeochem` from GitHub.

``` r
if (!require("devtools")) install.packages("devtools")
if (!require("rmarkdown")) install.packages("rmarkdown")
devtools::install_github("davidbajnai/isogeochem", build_vignettes = TRUE)
```

### Vignettes

Case studies demonstrating the use and scope of the functions in
`isogeochem` are available as vignettes.

``` r
browseVignettes("isogeochem")
```

## Dual clumped isotope thermometry

Use `D47()` and `D48()` to calculate equilibrium carbonate clumped
isotope values (*∆*<sub>47</sub>, *∆*<sub>48</sub>) for a given
temperature. `temp_D47()` calculates carbonate growth temperatures from
*∆*<sub>47</sub> values, while `temp_D48()` calculates growth
temperature corrected for kinetic effects considering both the
*∆*<sub>47</sub> and the *∆*<sub>48</sub> value.

``` r
if (!require("shades")) install.packages("shades")

# Model equilibrium carbonate ∆47 and ∆48 values
temp  = seq(0, 100, 10) # temperature range: 0—100 °C
D47eq = D47(temp, eq = "Fiebig21")
D48eq = D48(temp, eq = "Fiebig21")

# Sample data
D47_coral  = 0.617; D47_coral_err  = 0.006
D48_coral  = 0.139; D48_coral_err  = 0.022
D47_speleo = 0.546; D47_speleo_err  = 0.007
D48_speleo = 0.277; D48_speleo_err  = 0.029

## Plot in ∆47 vs ∆48 space ##
plot(0, type = "l", axes = TRUE, ylim = c(0.4, 0.7), xlim = c(0.1, 0.3),
     ylab = expression(Delta[47] * " (CDES90, ‰)"),
     xlab = expression(Delta[48] * " (CDES90, ‰)"),
     lty = 0, font = 1, cex.lab = 1, las = 1)

# Plot the equilibrium curve and points
lines (D48eq, D47eq, col = "purple", lwd = 2)
points(D48eq, D47eq, col = shades::gradient(c("blue", "red"), length(temp)),
       pch = 19, cex = 1.2)

# Plot the sample data,
# ... the kinetic slopes,
# ... and calculate growth temperatures corrected for kinetic effects
# ... using a single function!
temp_D48(D47_coral, D48_coral, D47_coral_err, D48_coral_err, ks = -0.6,
         add = TRUE, col = "seagreen", pch = 15)
#>   temp temp_err
#> 1   38        6
temp_D48(D47_speleo, D48_speleo, D47_speleo_err, D48_speleo_err, ks = -1,
         add = TRUE, col = "darkorange", pch = 17)
#>   temp temp_err
#> 1   30        9

# Add labels to the plot
text(D48(temp, eq = "Fiebig21"), D47(temp, eq = "Fiebig21"), paste(temp, "°C"),
     col = shades::gradient(c("blue", "red"), length(temp)), pos = 4, cex = 0.8)
```

<img src="man/figures/README-Figure1-1.png" width="60%" />

## Triple oxygen isotopes

`d17O_c()` calculates equilibrium carbonate oxygen isotope values
(*δ*<sup>18</sup>O, *δ*<sup>17</sup>O, *∆*<sup>17</sup>O) for a given
temperature and ambient water composition. Use the `mix_d17O()` function
to calculate mixing curves in triple oxygen isotope space, e.g., for
modeling diagenesis.

``` r
if (!require("shades")) install.packages("shades")

# Model equilibrium *calcite* precipitating from seawater
temp_sw  = seq(0, 50, 10) # temperature range: 0—50 °C
d18O_sw = 0 # d18O value of seawater
D17O_sw = -0.004 # D17O value of seawater
d18Op = prime(d17O_c(temp_sw, d18O_sw, D17O_sw, eq18 = "Daeron19")[, 1])
D17O  = d17O_c(temp_sw, d18O_sw, D17O_sw, eq17 = "Wostbrock20", eq18 = "Daeron19")[, 3]

# Model progressing meteoric diagenetic alteration
d18O_ds = -8 # d18O value of diagenetic fluid
D17O_ds = 0.020 # D17O value of diagenetic fluid
em_equi = d17O_c(temp = 10, d18O_H2O = d18O_sw, D17O_H2O = D17O_sw,
                 eq17 = "Wostbrock20", eq18 = "Daeron19") # equilibrium endmember
em_diag = d17O_c(temp = 80, d18O_H2O = d18O_ds, D17O_H2O = D17O_ds,
                 eq17 = "Wostbrock20", eq18 = "Daeron19") # diagenetic endmember
mix = mix_d17O(d18O_A = em_equi[1], d17O_A = em_equi[2],
               d18O_B = em_diag[1], d17O_B = em_diag[2], step = 25)

## Plot in ∆17O vs d'18O space ##
plot(0, type = "l", ylim = c(-0.1, 0.05), xlim = c(-10, 40), 
     xlab = expression(delta * "'" ^ 18 * "O (‰, VSMOW)"),
     ylab = expression(Delta ^ 17 * "O (‰, VSMOW)"),
     lty = 0, font = 1, cex.lab = 1, las = 1)

# Plot meteoric waters from the build-in dataset
points(prime(meteoric_water$d18O), D17O(meteoric_water$d18O, meteoric_water$d17O),
       col = "lightblue1", pch = 20)
text(-4, 0.05, "meteoric water", pos = 4, col = "lightblue1")

# Plot the composition of the fluids
points(prime(d18O_sw), D17O_sw, col = "darkmagenta", pch = 8) # seawater
text(prime(d18O_sw), D17O_sw, "seawater", pos = 4, col = "darkmagenta")
points(prime(d18O_ds), D17O_ds, col = "deeppink", pch = 8) # diagenetic fluid
text(prime(d18O_ds), D17O_ds, "diagenetic fluid", pos = 4, col = "deeppink")

# Plot the equilibrium curve and points
lines(d18Op, D17O, col = "darkmagenta", lwd = 2)
points(d18Op, D17O, pch = 19, cex = 1.2, 
       col = shades::gradient(c("blue", "red"), length(temp_sw)))
text(d18Op, D17O, paste(temp_sw, "°C"), pos = 4, cex = 0.8, 
     col = shades::gradient(c("blue", "red"), length(temp_sw)))
text(30, -0.05, paste("equilibrium calcite \nfrom seawater"),
     pos = 3, col = "darkmagenta")

# Plot the mixing model between the equilibrium and diagenetic endmembers
lines(prime(mix[, 1]), mix[, 2], col = "deeppink", lty = 3, lwd = 2)
points(prime(mix[, 1]), mix[, 2], pch = 18, cex = 1.5, 
       col = shades::gradient(c("#3300CC", "deeppink"), length(mix[, 2])))
text(prime(mix[, 1]), mix[, 2], paste(mix[, 3], "%", sep = ""), pos = 2, cex = 0.8, 
     col = shades::gradient(c("#3300CC", "deeppink"), length(mix[, 3])))
text(22, -0.09, paste("progressing", "\ndiagenetic alteration", "\n(recrystallisation) at 80°C", sep =""),
     pos = 2, col = "deeppink")
```

<img src="man/figures/README-Figure2-1.png" width="60%" />

## Thermometry

Use `isogeochem` to calculate crystallization temperatures from
carbonate *δ*<sup>18</sup>O and *∆*<sub>47</sub> values.

``` r
# Temperature from D47 with or without errors
temp_D47(D47_CDES90 = 0.601, eq = "Petersen19")
#> [1] 24.9
temp_D47(D47_CDES90 = 0.601,
         D47_error = 0.008 ,
         eq = "Anderson21")
#>   temp temp_err
#> 1 22.6      2.7

# Temperature from d18O
temp_d18O(
  d18O_c_VSMOW = 30,
  d18O_H2O_VSMOW = 0,
  min = "calcite",
  eq = "Watkins13")
#> [1] 25.9
```

## Fractionation factors

Use `isogeochem` to calculate <sup>16</sup>O/<sup>18</sup>O
fractionation factors at given temperatures.

``` r
if (!require("viridisLite")) install.packages("viridisLite")

plot(0, type = "l", las = 1, yaxt = "n", 
     xlim = c(10, 30), ylim = c(-30, 50),
     xlab = "Temperature (°C)",
     ylab = expression("Equilibrium enrichment in "^18*"O relative to H"[2]*"O (‰)"))
axis(2, seq(-30, 50, 10), las = 1)

temps = seq(10, 30, 1)
d18O_H2O_VSMOW = 0
cols = viridisLite::viridis(7, option = "C")

text(10, 45, expression("CO"[2]*" (aq)"), col = cols[1], adj = c(0, 0))
lines(temps, A_from_a(a18_CO2aq_H2O(temps), d18O_H2O_VSMOW),
      lwd = 2, lty = 2, col = cols[1])
text(10, 35, expression("HCO"[3]^"–"), col = cols[2], adj = c(0, 0))
lines(temps, A_from_a(a18_HCO3_H2O(temps), d18O_H2O_VSMOW),
      lwd = 2, lty = 2, col = cols[2])
text(10, 30, "calcite", col = cols[3], adj = c(0, 0))
lines(temps, A_from_a(a18_c_H2O(temps, "calcite", "Daeron19"), d18O_H2O_VSMOW),
  lwd = 2, lty = 1, col = cols[3])
text(10, 21, expression("CO"[3]^"2–"), col = cols[4], adj = c(0, 0))
lines(temps, A_from_a(a18_CO3_H2O(temps), d18O_H2O_VSMOW),
  lwd = 2, lty = 2, col = cols[4])
text(10, 1, expression("H"[2]*"O"), col = cols[5], adj = c(0, 0))
lines(temps, rep(d18O_H2O_VSMOW, length(temps)),
      lwd = 3, lty = 1, col = cols[5])
text(10, -23, expression("OH"^"–"), col = cols[6], adj = c(0, 0))
lines(temps, B_from_a(a18_H2O_OH(temps, eq = "Z20-X3LYP"), d18O_H2O_VSMOW),
  lwd = 2, lty = 1, col = cols[6])
```

<img src="man/figures/README-Figure3-1.png" width="60%" />

## Utility functions

``` r
# Convert between the VSMOW and VPDB scales:
to_VPDB(32)
#> [1] 1.05032
to_VSMOW(1)
#> [1] 31.95092

# Convert between classical delta and delta prime values:
prime(10)
#> [1] 9.950331
unprime(9.95)
#> [1] 9.999666

# Calculate isotope fractionation factors:
a_A_B(A = 30.40, B = 0.15)
#> [1] 1.030245
epsilon(a_A_B(A = 30.40, B = 0.15))
#> [1] 30.24546
```

## Datasets

Within `isogeochem` you have quick access to important datasets.

| Name             | Description                                                                    | Reference                               |
|------------------|--------------------------------------------------------------------------------|-----------------------------------------|
| `devilshole`     | The original Devils Hole carbonate *δ*<sup>18</sup>O time series               | Winograd et al. (2006)                  |
| `LR04`           | A benthic foraminifera *δ*<sup>18</sup>O stack                                 | Lisiecki & Raymo (2005)                 |
| `GTS2020`        | An abridged version of the GTS2020 oxygen isotope stack                        | Grossman & Joachimski (2020)            |
| `meteoric_water` | A compilation of meteoric water *δ*<sup>18</sup>O and *δ*<sup>17</sup>O values | Barkan & Luz (2010), Aron et al. (2021) |

For more information on the datasets please have a look at the
corresponding documentation, e.g., `?devilshole`

------------------------------------------------------------------------

## See also

There are several other R packages that complement `isogeochem` and are
worth checking out:

[`viridisLite`](https://github.com/sjmgarnier/viridisLite) and
[`viridis`](https://github.com/sjmgarnier/viridis) produce color-blind
and black-and-white printer friendly color scales.

[`clumpedr`](https://github.com/isoverse/clumpedr/) works with
[`isoreader`](https://github.com/isoverse/isoreader) to read in raw
measurement data and reproducibly process the results to clumped isotope
values.

[`seasonalclumped`](https://github.com/nielsjdewinter/seasonalclumped)
can be used to reconstruct temperature and salinity variations from
seasonal oxygen and clumped isotope records.

[`deeptime`](https://github.com/willgearty/deeptime) adds geological
timescales to ggplots.


NEWS.md
# isogeochem 1.1.1

## Function updates
* `d17O_c()` now allows to choose from two different 17O/18O fractionation factors.

## Minor improvements
* Reformatted `NEWS.md`

# isogeochem 1.1.0

[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.6861261.svg)](https://doi.org/10.5281/zenodo.6861261)

## Function updates
* `a18_c_H2O()` Can now calculate the dolomite/water 18O/16O fractionation factor.
* `a18_CO2acid_c()` Can now calculate the dolomite/CO2 18O/16O AFF.
* `a18...()` functions now return meaningful errors.
* `d17O_c()` Can now take into account a non-zero D17O value for the ambient water.
* `mix_d17O()` was overhauled. It now allows D17O values as input and specifying the resolution of the output. The output of the function now returns the d17O values as the fourth element in the data frame.

## New features
* `d17O_qz()` to calculate equilibrium oxygen isotope values for quartz.
* `meteoric_waters` is a compilation of d17O and d18O values for meteoric waters.

## Minor improvements
* Readme updates
* Vignette updates
* Description updates

# isogeochem 1.0.9

[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.5665966.svg)](https://doi.org/10.5281/zenodo.5665966)

## Minor improvements
* Vignette updates

# isogeochem 1.0.8

[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.5644934.svg)](https://doi.org/10.5281/zenodo.5644934)

## New features
* `D17O()` calculates triple oxygen isotope value from d18O and d17O values

## Function updates
* `york_plot()` returns silently

## Minor improvements
* Vignettes reset to the user's parameters and options
* Updated DOI links
* Readme updates
* Added webpage <https://davidbajnai.github.io/isogeochem/>

# isogeochem 1.0.7

[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.5552274.svg)](https://doi.org/10.5281/zenodo.5552274)

## New features
* `a18_CO2g_H2O()`
* `a18_CO2aq_H2O()`
* `a18_CO3_H2O()`
* `a18_HCO3_H2O()`
* `a13_CO2g_CO2aq()`
* `X_absorption()`
* `X_DIC()`

## Function updates
* In `temp_D48()` the curve_intersect() part was slimmed down
* Update to temp_D48() ii: temperatures are now calculated with the "Fiebig21" eq
* Added the "Anderson21" option to D47() and temp_D47()
* `temp_D47()` and `temp_D48()` both result in a similar data frame (length = 2) if errors are specified
* `temp_D47()`, `temp_D48()`, and `temp_d18O()` are `optimize`-d
* Added the "Kim07" eq to "aragonite" in `a18_c_H2O()`
* Added the "vanDijk18" eq to "siderite" in `a18_c_H2O()`
* Typo corrected in `a18_H2O_OH()`: "Z21-X3LYP" —> "Z20-X3LYP"

## Minor improvements
* Description updates
* Readme updates
* Vignette updates
* Installing `isogeochem` with vignettes now works (#2, @japhir)
* Improved and increased test coverage
* Updated logo

# isogeochem 1.0.1

[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.5507225.svg)](https://doi.org/10.5281/zenodo.5507225)

* This is the first stable release of isogeochem.


R/D48.R
# Functions in this file: D48(), temp_D48()

#' @title Carbonate D48 for a given temperature
#'
#' @description
#' `D48c()` calculates the equilibrium carbonate D48 value for a given temperature.
#'
#' @param temperature Carbonate growth temperature (°C).
#' @param eq Equation used for the calculation. Options are `"Fiebig21"` (default) and `"Swart21"`.
#'   `"Fiebig21"` refers to the CDES90 calibration in Fiebig et al. (2021).
#'   `"Swart21"` refers to the CDES90 "PBLM1" calibration in Swart et al. (2021).
#'
#' @return
#' Returns the carbonate D48 value expressed on the CDES90 scale (‰).
#'
#' @examples
#' D48c(33.7) # Returns 0.237
#' D48c(33.7,"Swart21") # Returns 0.239
#'
#' @references
#' Fiebig, J., Daëron, M., Bernecker, M., Guo, W., Schneider, G., Boch, R., et al. (2021).
#' Calibration of the dual clumped isotope thermometer for carbonates. Geochimica et Cosmochimica Acta.
#' <https://www.doi.org/10.1016/j.gca.2021.07.012>
#'
#' Swart, P. K., Lu, C., Moore, E., Smith, M., Murray, S. T., & Staudigel, P. T. (2021).
#' A calibration equation between D48 values of carbonate and temperature.
#' Rapid Communications in Mass Spectrometry, 35(17), e9147.
#' <https://www.doi.org/10.1002/rcm.9147>
#'
#' @family equilibrium_carbonate
#' @export

D48c = function(temperature, eq ="Fiebig21") {
  TinK = temperature+273.15
  if (eq == "Fiebig21") {
  1.028 *(6.002 / TinK - 1.299 * 10^4 / TinK^2 + 8.996 * 10^6 / TinK^3 - 7.423 * 10^8 / TinK^4) + 0.1245
  } else if (eq == "Swart21") {
  b = 0.088; m = 0.0142
  m * (10^6 / TinK^2) + b
  } else {
    warning("ERROR! Invalid input for eq")
  }

}

##————————————————————————————————————————————————————————————————————————————————##


#' @title Dual clumped isotope thermometry
#'
#' @description `temp_D48()` calculates carbonate growth temperature from D47 and D48 values.
#'
#' @param D47_CDES90 Carbonate D47 values expressed on the CDES90 scale (‰).
#' @param D48_CDES90 Carbonate D48 values expressed on the CDES90 scale (‰).
#' @param D47_error Error on the D47 value. Optional.
#' @param D48_error Error on the D48 value. Optional.
#' @param ks Kinetic slope. Has to be negative!
#' @param add Add graphics to an already existing plot? Default `FALSE`.
#' @param col Graphical parameter. Optional.
#' @param pch Graphical parameter. Optional.
#'
#' @return
#' Returns the carbonate growth temperature (‰).
#'
#' @examples
#' temp_D48(0.617, 0.139, ks = -0.6) # Returns 44
#' temp_D48(0.546, 0.277, ks = -1)   # Returns 33
#'
#' @references
#' References are listed at [D48c()] and [D47c()].
#'
#' @seealso [D47c()] calculates the equilibrium carbonate D47 value for a given temperature.
#' @seealso [D48c()] calculates the equilibrium carbonate D48 value for a given temperature.
#'
#' @details
#' The function calculates a D47 value as an intersect of two curves:
#' * the equilibrium D47 vs D48 curve from Fiebig et al. (2021)
#' * the kinetic slope
#'
#' The resulting D47 value is then converted to temperature using the [D47c()] function, i.e., by default the
#' equation of Petersen et al. (2019). This is not consistent and I will fix it in a later version.
#' In any case, the resulting discrepancy is smaller than the temperature error.
#'
#'@family thermometry
#'
#' @export

temp_D48 = function(D47_CDES90, D48_CDES90, D47_error, D48_error, ks, add = FALSE, col, pch) {
  line_sample = data.frame(x=c(D48_CDES90+1, D48_CDES90-1),
                           y=c(D47_CDES90+1*ks, D47_CDES90-1*ks))
  line_eq = data.frame(x=D48c(seq(-10,1000,0.1)), y=D47c(seq(-10,1000,0.1)))
  int = reconPlots::curve_intersect(line_sample, line_eq)

  temp = round(temp_D47(int$y),0)

  if (missing(D47_error) == FALSE && missing("D48_error") == FALSE) {
    line_cool = data.frame(x=c(D48_CDES90+D48_error+1, D48_CDES90+D48_error-1),
                           y=c(D47_CDES90+D47_error+1*ks, D47_CDES90+D47_error-1*ks))
    line_warm = data.frame(x=c(D48_CDES90-D48_error+1, D48_CDES90-D48_error-1),
                           y=c(D47_CDES90-D47_error+1*ks, D47_CDES90-D47_error-1*ks))
    int_cool = reconPlots::curve_intersect(line_cool, line_eq)
    int_warm = reconPlots::curve_intersect(line_warm, line_eq)
    temp_mean = temp
    temp_warm = round(temp_D47(int_warm$y),0)
    temp_cool = round(temp_D47(int_cool$y),0)
    temp = data.frame(temp_mean, temp_warm, temp_cool)

    if (add == TRUE) {
      if (missing(col) == T) col = "black" else col = col
      if (missing(pch) == T) pch = 19 else pch = pch
      graphics::rect(xleft = D48_CDES90-D48_error, ybottom = D47_CDES90-D47_error,
                     xright = D48_CDES90+D48_error, ytop = D47_CDES90+D47_error,
                     col = grDevices::adjustcolor(col, alpha.f = 0.3), border = NA)
      graphics::segments(D48_CDES90, D47_CDES90+D47_error, D48_CDES90, D47_CDES90-D47_error,
                         col="gray10", lwd=1, lty=1)
      graphics::segments(D48_CDES90-D48_error, D47_CDES90, D48_CDES90+D48_error, D47_CDES90,
                         col="gray10", lwd=1, lty=1)
      graphics::arrows(x0=D48_CDES90-D48_error, y0=D47_CDES90-D47_error,
                       x1=int_warm$x,           y1=int_warm$y,
                       code=0, col="gray70", lwd=1.5, lty=2)
      graphics::arrows(x0=D48_CDES90+D48_error, y0=D47_CDES90+D47_error,
                       x1=int_cool$x,           y1=int_cool$y,
                       code=0, col="gray70", lwd=1.5, lty=2)
    }
  }

  if (add == TRUE) {
    graphics::arrows(x0=D48_CDES90, y0=D47_CDES90,
                     x1=int$x,      y1=int$y,
                     code=2,  length = 0.15, col="black", lwd=1.5, lty=1)
    graphics::points(D48_CDES90, D47_CDES90,
           col = col, pch = pch, cex = 1.2)
  }

  return(temp)

}
